### PR TITLE
ENH: Exporting multiple nodes

### PR DESCRIPTION
**Pull request by ebrahimebrahim**
# Description

This PR addresses #6074 and extends the export functionality from #5876 by allowing multiple nodes to be exported in one batch. If you right click a a subject hierarchy item that has children, then "Export to file..." will allow you export the storable nodes that occur in the subtree rooted at your selection. You can choose to include children recursively, or to just include direct children. Multi-select is not supported here; only selecting a single subject hierarchy item with children can lead to the multiple-node-export functionality.

image

Each node _type_ occurring within the set of nodes slated for export will have its own options widget and its own export format dropdown box. 

image

The example in these screenshots does not show the "include children" or "recursively include children" checkboxes, because in this particular example:
- not including children would result in no nodes to export, so it is not an option
- the subtree rooted at the selected node has depth 1, so it doesn't make a difference whether you recursively include children or not-- hence the option is not displayed

So these options are only displayed when they matter. For example:

image

Notice that the filename entry textbox is disabled when multiple nodes are being exported. In that case, filenames are automatically set based on node name. The textbox is enabled when exporting a single node.

# Summary of changes
- The subject hierarchy export plugin has been modified to pass along lists of node IDs, rather than just a single node ID.
- The export dialog has been heavily reworked to handle lists of nodes. There is a checkbox for including children and a checkbox for doing so recursively. When nodes of multiple types are included, some widgets get created per node type.
- The export function in `qSlicerCoreIOManager` has been modified to accept a list of parameter sets, rather than just one parameter set. This allows us to pass to it multiple formats for different node types.
- A serious export-related bug (that I introduced) has been fixed where node IDs were sometimes being looked up in the main scene rather than in the temporary scene used for export.
- A mild export-related bug has been fixed where plane markups would get written by `vtkMRMLMarkupsJsonStorageNode` rather than `vtkMRMLMarkupsPlaneJsonStorageNode`, and similarly for ROIs. This caused them to be written incorrectly.

# Problems found

1. I found one case where exporting a transform fails silently, because success is returned by the storage node even when nothing is written. To see this happen, create a `vtkMRMLTransformNode` (not a subtype of it) and immediately try to export it, without modifying its parameters. Success is returned [because of these lines](https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLTransformStorageNode.cxx#L489-L493), whose intention I'm not understanding.

_This problem does not prevent this contribution from being merged._ It is non-critical and I think its a separate issue.

2. "Node type" currently means "class name". One possibly annoying consequence of this is that different types of markups (line, closed curve, etc.) are considered to be different node types, and therefore get their own export format dropdown and options widget, while in most cases the user would want the same settings for all of them. So when you go to export a folder that contains all your markups, you have to set your desired options separately for each markup subtype. Perhaps the same issue applies to different types of transforms. 

_This problem may need to be resolved_, depending on feedback as to how much of a problem it is. I'm looking for suggestions here. It would not be hard to redefine "node type" in my code, but I'm not sure what definition would make the most sense. It would be nice if there were a solution to this that doesn't treat markups as a special case.

### CODE REVIEW

File: Base/QTCore/qSlicerCoreIOManager.cxx
Comment (ebrahimebrahim) [inline Base/QTCore/qSlicerCoreIOManager.cxx:905]: This is the part that fixed the following bug I mentioned in my summary:
> A mild export-related bug has been fixed where plane markups would get written by `vtkMRMLMarkupsJsonStorageNode` rather than `vtkMRMLMarkupsPlaneJsonStorageNode`, and similarly for ROIs. This caused them to be written incorrectly.

**Comment by ebrahimebrahim** [comment]
@lassoan This is ready for a first review :grinning:

**Comment by ebrahimebrahim** [comment]
Notes from 2022.01.11 meeting:
- play with the layout so that we don't get the words "export format" and "options" repeated so many times
- for now maybe it is okay that different markups get their own widgets, since chosen options are at least preserved within the slicer session

**Comment by ebrahimebrahim** [comment]
@lassoan I adjusted the appearance to make it less noisy. Below is what it looks like now when there's a single node type vs multiple node types. Further suggestions are welcome!
export

**Comment by ebrahimebrahim** [comment]
When exporting multiple nodes, the title now shows the number, e.g. "Export 12 Nodes," rather than just "Export Multiple Nodes." Thanks @jcfr for the suggestion!

(this pr is ready for review any time)

**Comment by ebrahimebrahim** [comment]
@lassoan @pieper @jcfr @smrolfe @muratmaga
(just pinging some people to look for feedback and to indicate that this is ready for review :grinning:)

**Review by pieper — DISMISSED** [review]
@ebrahimebrahim since Jc plans to merge the python changes today let's hold off another day before finishing the merge.

**Comment by ebrahimebrahim** [comment]
@pieper Thank you for taking the time to review this!

I rebased to a more recent commit and gave this another few tests. I didn't have to change anything (besides doing a completely clean super-build after the python update).

@pieper @jcfr Are we good to merge now? Once the CI check is done.

**Comment by pieper** [comment]
Okay from my end.  @jcfr did you learn everything you need about the python integration so that we are open to merging other PRs?

**Pull request merged**
